### PR TITLE
Add support for  Lua packages

### LINKED ISSUE #654: More math functions for mta
**Is your feature request related to a problem? Please describe.**
Low amount of math function that are used often

**Describe the solution you'd like**
Add more math functions

**Describe alternatives you've considered**
Copy and paste one piece of code to every resource

**Additional context**
few math functions that could be added, all of them you can find on wiki and in google getDistanceBetweenPointAndSegment2D  getDistanceBetweenPointAndSegment3D, findRotation2D, findRotation3D, getPointFromDistanceRotation, getOffsetFromXYZ
except that, should be added function for bazier curve and splines in 2d and 3d

### PR DESCRIPTION
**Pull request by qaisjp**
Work done by @darkdreamingdan.

This pull request is to record any remaining hurdles that need to be jumped before merging.

Personally, I am only happy with merging this if it's _forward compatible_ with an expansion of this feature to support [LuaRocks](https://luarocks.org). Discussion about LuaRocks should be done in a separate issue.

Fixes #654

### CODE REVIEW

**Comment by darkdreamingdan** [comment]
I think this should work with LuaRocks

**Comment by CrosRoad95** [comment]
make `require` allow get lua files outside mta folder which is defined, example /var/mta on linux
this could be useful for hostings, could decrease amount of files and speed up scripting

hostings can control mtaserver.conf file and keep this path always in place, don't let user to change it
for security, mta while running could checking this path and if contain file like mta.txt with content `allow this folder for require`, if not, then this folder will not work for `require`

File: Shared/mods/deathmatch/logic/luadefs/CLuaUtilDefs.cpp [hunk 1]
```diff
@@ -716,3 +719,40 @@ int CLuaUtilDefs::tocolor(lua_State* luaVM)
...
+        if (lua_type(luaVM, -1) != LUA_TNIL)
+            return 1;
+        lua_pop(luaVM, 1);            // stack: ["moduleName"]
+
+        // Check whether the appropriate pure lua module exists
+        if (pLuaMain->LoadLuaLib(luaVM, strMod, strError))            // stack: ["moduleName",pkgLuaMod/nil]
+            return 1;
+#ifndef MTA_CLIENT
+        strError += "\n";
+        // Check if a C library exists
```
Comment (CrosRoad95) [inline Shared/mods/deathmatch/logic/luadefs/CLuaUtilDefs.cpp:749]: libraries in c/c++ will allowed?

**Comment by CrosRoad95** [comment]
I made some tests, serverside ( shared code, i don't think there are difference between client and server ).

### Checks for incorrect behaviour
1. simply `require("module")`
i think, should show only path from mta root directory
image

2. while require myself filename = "s.lua" and `require "s"`. I don't think this is issue, but could information `you can't require myself`
image

3. I created 3 files `text.lua` in each of possible according to pattern:
`print("test in ", "D:/mtasa require test/InstallFiles/server/mods/deathmatch/resources/t")`
where second argument is current pattern ( from screenshot above ) and properly loaded only first occur 

4.
```lua
longText = string.rep("a",2^24)
require(longText) -- just spam in console, and a bit lag
```

5.
```lua
require("../../../../../asdf")
```
image

6.
i was suprised when i discovered this
```lua
require("ŕşôíŠ%v=Ü6ÇŘ˙˙śeDL×`­ëGF–±Öş|5ľxuóľ") -- https://pastebin.com/eA1xHZQr
```
when you trying require this monster, windows produce this sound:
https://www.youtube.com/watch?v=qlUFWSiOXpM

7.
```lua
setTimer(function()
  for k=1,100 do
    randomString = {}
    for i=1,40 do
    randomString[#randomString + 1] = string.char ( math.random(255) )
    end
    randomString = table.concat(randomString)
    require(randomString)
  end
end,50,0)
```
nothing special, just make server logs spam
i ran this for few minutes, crash doesn't appear, server working normally after stopping resource

8. Test while required script has error
image

9. Creating elements in required resource working. I restarted resource many times, and always print same thing

`test.lua`
```lua
a = createElement("test")
return a
```
```lua
rrr = require "test"

iprint(rrr, getElementsByType("test"))
```
image

### Tests in practice usage
1.
```lua
rrr = require "test"

iprint(rrr, rrr.func(2,2))
```
where `test` is
```lua
randomModule = {}

randomModule.int = 1
randomModule.float = math.pi
randomModule.string = "hi"
randomModule.table = {a = 3, [1] = 9, asdf = "asdf"}
randomModule.func = function(a,b) return a+b end

return randomModule
```
image

2. Traceback working

`test.lua`
```lua
function traceback ()
  local level = 1
  while true do
    local info = debug.getinfo(level, "Sl")
    if not info then break end
    if info.what == "C" then   -- is a C function?
      print(level, "C function")
    else   -- a Lua function
      print(string.format("[%s]:%d",
                          info.short_src, info.currentline))
    end
    level = level + 1
  end
end

return traceback
```
`t.lua`
```lua
rrr = require "test"

iprint(rrr())
```

image


### Bugs, issues

1. Infinity loop wrong source information

`s.lua`
```lua
print ("blabla")

rrr = require "test" -- contain infinity loop

print("bla bla")
```
image

2. require myself should be disabled.

3. While incorrect file is passed, only path from server root should be showned #683

File: Shared/mods/deathmatch/logic/luadefs/CLuaUtilDefs.cpp (hunk 1 above)
Comment (qaisjp) [reply Shared/mods/deathmatch/logic/luadefs/CLuaUtilDefs.cpp:749]: Yes. See commit message of https://github.com/multitheftauto/mtasa-blue/commit/a59316354482ee89687a95d238ee0078ae006a6d

Put behind ACL too. Server only.

File: Server/mods/deathmatch/logic/lua/CLuaMain.cpp [hunk 2]
```diff
@@ -658,3 +660,191 @@ int CLuaMain::OnUndump(const char* p, size_t n)
...
+    SString strPath = strName;
+    // Name format shouldn't include slashes.  Subdirs are dots.
+    ReplaceOccurrencesInString(strPath, "\\", "");
+    ReplaceOccurrencesInString(strPath, "/", "");
+    ReplaceOccurrencesInString(strPath, ".", "/");
+
+    strPath = PathJoin(g_pServerInterface->GetModManager()->GetServerPath(), SERVER_BIN_PATH_MOD, "modules", strPath);
+
+#ifdef WIN32
+    strPath += ".dll";
```
Comment (CrosRoad95) [inline Server/mods/deathmatch/logic/lua/CLuaMain.cpp:685]: or `.lua`

Comment (darkdreamingdan) [reply Server/mods/deathmatch/logic/lua/CLuaMain.cpp:685]: No, `.lua` is available in `LoadLuaLib`

Comment (CrosRoad95) [reply Server/mods/deathmatch/logic/lua/CLuaMain.cpp:685]: i just tested, and lua script with .dll extension not working

Comment (darkdreamingdan) [reply Server/mods/deathmatch/logic/lua/CLuaMain.cpp:685]: What do you mean? There is a separate function, `LoadLuaLib`, for loading .lua files.  That is already working.  This function is for loading C extensions.

**Review by patrikjuvonen — CHANGES_REQUESTED** [review]
Summary of what needs fixing and/or double-checking:
  - [ ] Infinite loop error should output some more useful trace
  - [x] Error out `require()` if trying to require the same script file

Other than that seems fine with #683 merged into this.
I will test LuaRocks with this soon and add any further notes on this review.

**Comment by CrosRoad95** [comment]
any plans to marge?

**Comment by qaisjp** [comment]
@patrikjuvonen — I have investigated requiring the same file.

Test code:

```
s = io.read("*l")
if s == "yes" then
    print("YUP")
    require("test")
end
print(s, s:len())
```

- Lua 5.1 (MTA's Lua, but not tested in MTA) errors out after the second time you enter "yes" (so this file is run a total of two files, once on initial load, and once after the first require)
- Lua 5.3.5 simply repeatedly says "YUP" (after entering "yes")

I think we should just do the latter.

- [ll_require in Lua 5.1](https://github.com/multitheftauto/mtasa-blue/blob/28bc42e/vendor/lua/src/loadlib.c#L451)
- [ll_require in Lua 5.3.5](https://github.com/lua/lua/blob/f995095/loadlib.c#L613)

Here is the [commit in Lua](https://github.com/lua/lua/commit/0a6a6b9d9d7b4ad0d627a8841ce4409e437bf70e) that removes this error message.

The Lua specification also does not define what should happen if `require` requires itself.

**Comment by qaisjp** [comment]
Also, this is how 5.3.5 behaves for infinite loops:

```lua
➜ cat test.lua
require "test"
print "hello"

➜ lua test.lua
lua: C stack overflow
stack traceback:
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	...
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	./test.lua:1: in main chunk
	[C]: in function 'require'
	test.lua:1: in main chunk
	[C]: in ?
```

**Comment by patrikjuvonen** [comment]
Right now requiring the same file crashes the server.

**Comment by buepas** [comment]
Since the development of this PR has halted (has it?), I think it helps if there is an up-to-date todolist of what has to be done in order to complete this PR. 
To-dos so far:
* More detailed error messages for Pure lua packages (?)
* Extensive testing
* Infinite loop error should output some more useful trace

**Comment by patrikjuvonen** [comment]
A new PR should be created by someone who wants to author this enhancement to the end. Remember to link to this, gather necessary information from here and elsewhere to the new PR.

**Pull request closed without merging**